### PR TITLE
RoR - Private events associations project

### PR DESCRIPTION
**Pull request by guirecordon**
### Collaborative project - Request for review
**We did the following for this project and now feel ready to request for a review. We: **

- created a User, an Event, and an EventManager model
- wrote the associations for them with their foreign keys and many-to-many relationships
- wrote validations for the users and event creators to be able to log in and only have access to their respective actions
- created a way for the users to be able to accept their invitations once they were made by another user. We did that by adding a 'confirmation' column to our EventManager table and only displaying the events once they were marked as true. They were all set to false in the beginning.
- we wrote integration tests and unit tests for authentication and models

thank you.

### CODE REVIEW

**Review by mosesogwo — CHANGES_REQUESTED** [review]
**Hello!:wave: Good start on this project:+1: Here are your required improvements:**

:point_right: Event's index page should display past and upcoming events lists. This isn't working right now. Events are not being displayed.



:point_right: The event's show page displays the list of attendants. Attendants are not  being displayed currently.



:point_right:  The user's show page displays events the user has created and events that the user is attending to separated in past and upcoming. The events are not visible on the page.

:point_right: Event model has an event date attribute. So you have to add the date attribute to your event creation form.



:point_right: Events controller index action should list past and upcoming events. This grouping should be made in the controller and not views.

;point_right: Users controller show action should have the following:

```
lists events created by user, i.e. @created_events =  current_user.events (screenshot from Odin)

 lists events previoulsy attendented by user using event model scope, i.e. @past_events = current_user.attended_events.previous (screenshot from Odin)

 lists upcoming events previously attended by user using event model scope, i.e. @upcoming_events = current_user.attended_events.upcoming (screenshot from Odin)

```
:point_right: Event management feature spec should cover the entire event management feature. You should send an invite in your spec and verify that it's working.

:point_right: Your tests are failing. Please fix them.

_**Kindly make these changes and submit another code review request.**_

**Any questions? You can reach me on Slack (Moses Ogwo)**

**Cheers:stars:**

### Status: Changes Requested:exclamation:

**Comment by guirecordon** [comment]
Thank you, Moses.
We worked on the changes you requested. 

- The upcoming and past events should be working now
- Attendants of the events should also be displayed (we had to change the conditional for the invite accept in the users controller -- confirmed.zero?)
- The date attribute is working and we also added a couple of links, etc for better user experience.
- Finally, we seem to have been able to get the EventManager RSpec test to pass, but that was a tricky one. So we'd love to get some feedback on it.

We're requesting a new review.
Thank you

File: app/views/users/show.html.erb [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+      <p><em> You got no pending invitations </em> </p>
+    <% else %>
+      <% @pending_invitations.each do |e| %>
+        <li>
+          <p>Invitation ID: <%= link_to e.attended_event_id, event_path(e.attended_event_id) %></p>
+          <p><%#= e.name %></p>
+          <p><%= link_to "Accept Invitation", event_manager_path(e) %></p>
+        </li>
+      <% end %> 
+    <% end %>
```
Comment (maelfosso) [inline app/views/users/show.html.erb]: Kindly remove login from views, move it in view helpers

File: app/views/layouts/application.html.erb [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+      <div class="nav-bar d-flex">
+        <p>Logged in as <%= current_user.username %></p>
+        <%= link_to "Log Out", session_path(current_user), method: :delete %>
+      </div>  
+    <% else %>
+      <div class="nav-bar d-flex">
+        <%= link_to "Sign up", new_user_path %> or
+        <%= link_to "Log In", new_session_path %>
+      </div>
+    <% end %>
```
Comment (maelfosso) [inline app/views/layouts/application.html.erb]: Move logic into view helpers

File: app/views/events/index.html.erb [hunk 3]
```diff
@@ -0,0 +1,24 @@
+<h1> Events</h1>
+
+<% if current_user %>
+  <%= link_to "Create New Event", new_event_path %>
+<% end %>
```
Comment (maelfosso) [inline app/views/events/index.html.erb]: Move logic into view helpers

File: app/views/events/show.html.erb [hunk 4]
```diff
@@ -0,0 +1,18 @@
+<h1><%= @event.name %></h1>
+<p>Event created by <%= @event.creator.username %></p>
+<h2>Invited Guests</h2>
+    <% if @attendants.size == 0 %>
+        <p><em>No guests have been invited to this event so far.</em></p>
+    <% else %>
+        <% @attendants.each do |e| %>
+        <%= e.username %><br>
+    <% end %>
```
Comment (maelfosso) [inline app/views/events/show.html.erb]: Move logic into view helpers

File: app/controllers/users_controller.rb [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+      flash.alert = 'fields cannot be blank!'
+      render 'new'
+    end
+  end
+
+  def show
+    @user = User.find(params[:id])
+    @pending_invitations = []
+    @upcoming_events = current_user.attended_events.where('date >= ?', DateTime.now)
+    @previous_events = current_user.attended_events.where('date <= ?', DateTime.now)
```
Comment (maelfosso) [inline app/controllers/users_controller.rb]: No longer need to add request. You already have the scope `Event`.
Directly used
```rb
@upcoming_events = current_user.attended_events.upcoming
@previous_events = current_user.attended_events.past
```

File: app/controllers/users_controller.rb [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+
+    @user.event_managers.each do |invite|
+      confirmed = invite.confirmation
+      @pending_invitations << invite if confirmed.zero?
+    end
```
Comment (maelfosso) [inline app/controllers/users_controller.rb]: No need to confirm the invitation!

**Review by maelfosso — CHANGES_REQUESTED** [review]
# Changes required :red_circle: 

Hello,

Thank you for applying the suggested changes
However, there are still some issues 

1. There are no information concerning the attended users, as required [here](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/events_attendees_requirement.png).
Screenshot at 2020-09-14 19:09:50
But if you take a look at the database, you will notice that there are invited users
```rb
2.7.1 :004 > EventManager.all
  EventManager Load (0.3ms)  SELECT  "event_managers".* FROM "event_managers" LIMIT ?  [["LIMIT", 11]]
 => #<ActiveRecord::Relation [#<EventManager id: 1, attendee_id: 1, attended_event_id: 1, created_at: "2020-09-14 18:02:40", updated_at: "2020-09-14 18:02:40", confirmation: 0>, #<EventManager id: 2, attendee_id: 1, attended_event_id: 4, created_at: "2020-09-14 18:04:00", updated_at: "2020-09-14 18:04:00", confirmation: 0>]> 
2.7.1 :005 > Event.all
  Event Load (0.3ms)  SELECT  "events".* FROM "events" LIMIT ?  [["LIMIT", 11]]
 => #<ActiveRecord::Relation [#<Event id: 1, name: "Music go", date: "2020-09-17 00:00:00", created_at: "2020-09-14 18:02:31", updated_at: "2020-09-14 18:02:31", creator_id: 1, description: "Yeah you are OK. https://github.com/emasdev/privat...">, #<Event id: 2, name: "Music go", date: "2020-09-01 00:00:00", created_at: "2020-09-14 18:02:50", updated_at: "2020-09-14 18:02:50", creator_id: 1, description: "Jour de ca ah bon!!!">, #<Event id: 3, name: "Pourrational high", date: "2020-09-30 00:00:00", created_at: "2020-09-14 18:03:22", updated_at: "2020-09-14 18:03:22", creator_id: 1, description: "Yeah you are OK. https://github.com/emasdev/privat...">, #<Event id: 4, name: "Reviewing", date: "2020-10-03 00:00:00", created_at: "2020-09-14 18:03:51", updated_at: "2020-09-14 18:03:51", creator_id: 2, description: "Read it correctly youre chaongre ;w;oijwe">]> 
```

2- There is no link to access to the user show page which is requested. [Check it here](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/user_events_requirement.png)

3- The model `AddCreatorToEvent`  is never used. Can you remove it ?

4- Read the comments below

Kindly make the suggested changes and submit a new code review request
Thanks

Mael FOSSO

**Comment by guirecordon** [comment]
Thank you, Moses.

We removed all the logic from views and moved them to helpers; added the links you requested and made the other changes.

Thank you.

**Review by Dandush03 — APPROVED** [review]
# Status: :rocket: Approved :rocket: :smile: 
# Keep it up
# :computer: Happy Coding :computer:

**Pull request merged**